base/cvd/cuttlefish/host/frontend/webrtc/audio_mixer.cpp

```diff
@@ -0,0 +1,278 @@
+  mixer_cv_.notify_one();
+  if (mixer_thread_.joinable()) {
+    mixer_thread_.join();
+  }
+}
+
+void AudioMixer::OnStreamStopped(uint32_t stream_id) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  next_frame_.erase(stream_id);
+}
+
+void AudioMixer::OnPlayback(uint32_t stream_id, uint32_t stream_sample_rate,
+                              uint8_t stream_channels_count,
+                              uint8_t stream_bits_per_channel,
+                              const uint8_t* buffer, size_t size) {
+  static const std::vector<std::vector<float>> kDirectMap{{
+      {1, 0, 0, 0, 0, 0},
+      {0, 1, 0, 0, 0, 0},
+      {0, 0, 1, 0, 0, 0},
+      {0, 0, 0, 1, 0, 0},
+      {0, 0, 0, 0, 1, 0},
+      {0, 0, 0, 0, 0, 1},
+  }};
+
+  const auto stream_frames_count =
+      GetFramesCount(size, stream_channels_count, stream_bits_per_channel);
+  const auto frames_count = GetFrameCountAfterResampling(
+      sample_rate_, stream_sample_rate, stream_frames_count);
+
+  std::unique_lock<std::mutex> lock(mutex_);
+  const bool need_notify = next_frame_.empty();  // no active streams
+
+  const auto it = next_frame_.find(stream_id);
+  // If stream was not active it will start from the next 10ms bucket
+  const auto next_frame_id = it == next_frame_.end() ? 0 : it->second;
+
+  // Resize the buffer to fit extra frames_count samples
+  const auto buffer_size = mixed_buffer_.size();
+  if (buffer_size < (next_frame_id + frames_count) * frame_size_bytes_) {
+    mixed_buffer_.resize((next_frame_id + frames_count) * frame_size_bytes_);
+  }
+
+  if (next_frame_id + frames_count > last_active_frame_) {
+    std::fill(mixed_buffer_.data() + last_active_frame_ * frame_size_bytes_,
+              mixed_buffer_.data() +
+                  (next_frame_id + frames_count) * frame_size_bytes_,
+              0);
+  }
+
+  const auto stream_sample_size_bytes = stream_bits_per_channel / 8;
+  const auto convert_fn =
+      kConvertAudioStreamFunctionMap[sample_size_bytes_]
+                                    [stream_sample_size_bytes];
+  CHECK(convert_fn) << "Format is not supported";
+  const auto filled_frames_count =
+      convert_fn(mixed_buffer_.data() + next_frame_id * frame_size_bytes_,
+                 channels_count_, sample_rate_, buffer, stream_channels_count,
+                 stream_sample_rate, stream_frames_count, kDirectMap);
+  CHECK(filled_frames_count <= frames_count);
+
+  next_frame_[stream_id] = next_frame_id + filled_frames_count;
+  last_active_frame_ =
+      std::max(last_active_frame_, next_frame_id + filled_frames_count);
+
+  if (need_notify) {
+    lock.unlock();
+    mixer_cv_.notify_one();
+  }
+}
+
+void AudioMixer::MixerLoop() {
+  static constexpr auto kInterval = 10;  // milliseconds
+
+  int64_t next_frame_time = rtc::TimeMillis();
+
+  while (!stop_mixer_.load()) {
+    const auto delay = next_frame_time - rtc::TimeMillis();
+    if (delay > 0) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(delay));
```

Comment:
Use `std::condition_variable::wait_until` (or `wait_for`) so that it can be interrupted by a stop.